File: puente/apps/mobile/app/src/main/java/ai/puente/core/SessionState.kt
package ai.puente.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.encodeToJsonElement

/** State machine de la app (puente/CLAUDE.md §12). */
enum class PuenteSessionState {
    DISCONNECTED,
    CONNECTED_IDLE,
    LISTENING,           // PTT
    PROCESSING,          // llamada al worker
    SPEAKING,            // TTS sonando
    SENTIDO_CONTINUOUS,  // 1 frame / N s
    ESCANEO_LIVE,        // Gemini Live
}

@Serializable
data class ShoppingItem(
    val item: String,
    var status: String = "pending",      // pending | done
    val preferencia: String? = null,
)

/**
 * SessionState que viaja a /agents/super (schemas.md → SessionState).
 * La app es dueña de la lista; Hermes la actualiza y devuelve la versión nueva.
 */
@Serializable
data class SessionState(
    val session_id: String,
    val usuario_id: String,
    val super_id: String,
    var visita_numero: Int = 1,
    val lista_compra: MutableList<ShoppingItem> = mutableListOf(),
    var ubicacion_estimada: String? = null,
    val items_en_carrito: MutableList<String> = mutableListOf(),
    var turno_actual: String? = null,
    var pending_confirm: Boolean = false,
) {
    fun pendientes(): List<ShoppingItem> = lista_compra.filter { it.status == "pending" }

    fun toJson(json: Json): JsonObject =
        json.encodeToJsonElement(serializer(), this) as JsonObject

    companion object {
        fun fromJson(json: Json, obj: JsonObject): SessionState =
            json.decodeFromJsonElement(serializer(), obj)
    }
}

File: puente/apps/mobile-ios/Puente/Core/SessionState.swift
import Foundation

enum PuenteSessionState: String {
    case disconnected = "DISCONNECTED"
    case connectedIdle = "CONNECTED_IDLE"
    case listening = "LISTENING"
    case processing = "PROCESSING"
    case speaking = "SPEAKING"
    case sentidoContinuous = "SENTIDO_CONTINUOUS"
}

struct ShoppingItem: Codable {
    var item: String
    var status: String
    var preferencia: String?
}

struct SessionState: Codable {
    var sessionId: String
    var usuarioId: String
    var superId: String
    var visitaNumero: Int
    var listaCompra: [ShoppingItem]
    var ubicacionEstimada: String?
    var itemsEnCarrito: [String]
    var turnoActual: String?
    var pendingConfirm: Bool

    enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case usuarioId = "usuario_id"
        case superId = "super_id"
        case visitaNumero = "visita_numero"
        case listaCompra = "lista_compra"
        case ubicacionEstimada = "ubicacion_estimada"
        case itemsEnCarrito = "items_en_carrito"
        case turnoActual = "turno_actual"
        case pendingConfirm = "pending_confirm"
    }

    func asDictionary() -> [String: AnyEncodable] {
        guard let data = try? JSONEncoder().encode(self),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return obj.mapValues { AnyEncodable($0) }
    }
}

func pendientes(_ s: SessionState) -> [ShoppingItem] {
    s.listaCompra.filter { $0.status == "pending" }
}

func applyAgentState(_ s: inout SessionState, updated: [String: AnyDecodable], pending: Bool) {
    if let list = updated["lista_compra"]?.value as? [[String: Any]] {
        s.listaCompra = list.compactMap { dict in
            guard let item = dict["item"] as? String else { return nil }
            let status = dict["status"] as? String ?? "pending"
            return ShoppingItem(item: item, status: status, preferencia: dict["preferencia"] as? String)
        }
    }
    if let cart = updated["items_en_carrito"]?.value as? [String] {
        s.itemsEnCarrito = cart
    }
    if updated["ubicacion_estimada"] != nil {
        s.ubicacionEstimada = updated["ubicacion_estimada"]?.value as? String
    }
    if updated["turno_actual"] != nil {
        s.turnoActual = updated["turno_actual"]?.value as? String
    }
    s.pendingConfirm = pending
}
